### ISSUE TITLE
New Rule: Require specific font size units

### ISSUE THREAD

**Issue by nasty23-star** [eslint/css#73]
### Environment

Rule details
Flag the use of px

What type of rule is this?
Warns about a potential problem

Example code
.card {
    font-size: 12px;
}

Participation

I am willing to submit a pull request to implement this rule.
Additional comments
Needing !important indicates there may be a larger underlying issue.

### What problem do you want to solve?

Rem and em are more accessibility-friendly than px.

### What do you think is the correct solution?

The option to enable accessibility mode and warn about usage of px.

### Participation

- [ ] I am willing to submit a pull request for this change.

### Additional comments

I think it's very useful for improving accessibility.
https://www.a11y-collective.com/blog/what-is-rem-in-css/

**Comment by nzakas** [comment]
Thanks for the suggestion. This seems a bit too specific to be included as a general rule. (We typically don't create rules that simply disallow specific syntax.) 

I think we'd be open to a rule that was a bit broader, maybe something that encouraged the use of specific font size types? So rather than just disallowing `px`, maybe it would look for a small set of "good" size types?

**Comment by nasty23-star** [comment]
Thank you for the clarification. Your suggestion sounds very good.

**Comment by Tanujkanti4441** [comment]
>I think we'd be open to a rule that was a bit broader, maybe something that encouraged the use of specific font size types? So rather than just disallowing px, maybe it would look for a small set of "good" size types?

How about a rule called `required-font-unit` or `required-font-size-unit` that will take a set of strings such as `rem`, `px` etc. as option, so rule would allow only these units for font-size?

**Comment by nzakas** [comment]
> How about a rule called `required-font-unit` or `required-font-size-unit` that will take a set of strings such as `rem`, `px` etc. as option, so rule would allow only these units for font-size?

I think this is heading in the correct direction. `require-font-size-unit` (without the "d" in require) seems like it would make the most sense as a name, although I wonder if this rule needs an allow list or a deny list?

**Comment by Tanujkanti4441** [comment]
> although I wonder if this rule needs an allow list or a deny list?

I think for consistency, using an allowlist would be a good idea?

**Comment by nzakas** [comment]
Makes sense. I'd like to find an article or something online with recommendations for which font size units are considered best practices so we can have a logical default setting.

**Comment by Tanujkanti4441** [comment]
Sure! these are some articles that suggest `rem` for font size:

>It's a good idea to use `rem` to set font sizes because it is designed to adapt to the user's browser preferences. This helps with accessibility. It is also good for consistent scaling, as changing the HTML font size will affect the elements with rem units.

https://www.freecodecamp.org/news/css-units-when-to-use-each-one/

> Tip: The `em` and rem units are practical in creating perfectly scalable layout!

https://www.w3schools.com/css/css_units.asp

> Set Font Sizes with rem: Use `rem` units for setting font sizes in most cases. rem ensures that font sizes are relative to the root element's font size, making them responsive to user preferences and zooming.

https://medium.com/design-bootcamp/css-font-sizing-beyond-pixels-for-better-design-35457c2650d3

**Comment by nzakas** [comment]
Thanks! Going to post on social media to see what other responses we get.

**Comment by bradzacher** [comment]
The thing we found while working with typescript-eslint is that almost any time you create a `no-` prefixed rule, someone will want to go against the "no" and instead enforce the inverse.

I'd personally suggest a name like `consistent-font-sizes` which would then give you room to add an option for each style of font size whilst still adding a default value to enforce the perceived best practice.

**Comment by ericfortis** [comment]
The rule would have to be global (not font-specific).

For example, this would break design proportions.
```css
button {
  font-size: 1rem;
  padding: 4px;
}
```

For not breaking it, you'd have to do this, which negates the a11y point of allowing users to set a custom browser font size.
```css
html { font-size: 16px }
```
---

At any rate, If you stumbled upon this rule, and you are not a designer, use the units your designers use. Otherwise, you'll have to compute a lot. For example, the designer says `540px` and you’d have to write `33.75rem`

---

Q. **Do vision impaired users set a custom font size?** I'd guess they use low-resolution and the browser zoom, given that most web pages are in px.

**Comment by nzakas** [comment]
@ericfortis I assume by "global" you mean it would apply to more than just `font-size`?

I think that's beyond the scope of what we're going for with this rule. The point isn't to ensure an entire UI scales together, it's to enforce a requirement for font size units.

**Comment by nzakas** [comment]
After digging into this a bit, I'm wondering if what we really want is a `relative-font-units` rule that enforces the use of relative units?

Let's say by default it enforces the use of `rem`, but you can optionally specify other units to allow, like this:

```js
{
    "relative-font-units": ["error", { allow: ["rem", "em", "%" ] }]
}
```

Here's a list of relative units I found:

*   **cap**: Represents the "cap height" (nominal height of capital letters) of the element's font.
*   **ch**: Represents the width or advance measure of the "0" glyph in the element's font.
*   **em**: Represents the calculated font-size of the element.
*   **ex**: Represents the x-height of the element's font.
*   **ic**: Equal to the advance measure of the "水" glyph (CJK water ideograph) in the font.
*   **lh**: Equal to the computed line-height of the element.
*   **rcap**: Equal to the "cap height" of the root element's font.
*   **rch**: Equal to the width or advance measure of the "0" glyph in the root element's font.
*   **rem**: Represents the font-size of the root element.
*   **rex**: Represents the x-height of the root element's font.
*   **ric**: Equal to the `ic` unit on the root element's font.
*   **rlh**: Equal to the `lh` unit on the root element's font.

We could limit the options in `allow` to just those known relative units to ensure that people don't accidentally include absolute units.

Thoughts?

**Comment by kylemh** [comment]
[This is a really well-written article](https://www.joshwcomeau.com/css/surprising-truth-about-pixels-and-accessibility/) about relative units of CSS and when to use them from the perspective of accessibility and maintainability from @joshwcomeau.

I don't think all relative units should be acceptable. I think specifically and exclusively `rem`s should be allowed, at least with the default. If the mindset is to improve accessibility, allowing % or em units are much more fragile to implementation details. During development, a dev could take a UI from accessible to not accessible without knowing it if units with parent-relative values are used. With `rem`, they'd be affecting the entire site in - hopefully - a more obvious manner.

**Comment by illarionvk** [comment]
Consistent units across the project, not only font-size, would be more important for my team. 

For example, we require all px values to be wrapped in a `rem-calc()` Sass function.

We also use em values when it makes sense to inherit parent value

**Comment by nzakas** [comment]
@illarionvk that would be beyond the scope of the type of rule we're looking to add here. You could definitely create a custom rule that does what you're asking.

**Comment by Tanujkanti4441** [comment]
I am in favor of `relative-font-units`, would like to implement this if accepted.

**Comment by nzakas** [comment]
Marking as accepted. @Tanujkanti4441 you got it!
